Все задачи с Codewars + тесты в отдельной директории + улучшение кода (теперь мой код на вычисление алгебраической функции умеет работать с аргументами командной строки)

EugeneWWolf opened the pull request:
Группа 1/147

Edit 30.11: Теперь мой код работает с аргументами запуска. Если ничего не вводить, то программа запустится со стандартными значениями. Аргументы запуска можно увидеть, напечатав `go run main.go -h`

Edit 02.12: Залила в отдельный файл все решения задач с Codewars, которые вы присылали в дискорде

EugeneWWolf committed:
Added tests for function, which evaluates algebraic function's value

EugeneWWolf committed:
Moved tests to separete directory, moved code which I need to test to different from main package. Changed module name in go.mod from "isuct.ru/informatics2022" to "myProject" and etc.

jskonst commented on golang/go.mod:
```diff
@@ -1,3 +1,3 @@
-module isuct.ru/informatics2022
+module myProject
```
вот это в целом не влияет - просто отсчет ведется от `isuct.ru/informatics2022`
и то что у вас стало `myProject/internal` должно было бы быть `isuct.ru/informatics2022/internal`

jskonst commented on golang/internal/functions.go:
```diff
@@ -0,0 +1,15 @@
+	return functionValue
+}
+
+func PrintFunctionValue(x, functionValue float64) {
+	fmt.Printf("f(%.3g) = %.13g\n", x, functionValue)
+}
```
Это уж какая-то микрооптимизация :) смысла нет

jskonst commented on golang/tests/functions_test.go:
```diff
@@ -0,0 +1,34 @@
+	"myProject/internal"
+)
+
+func TestEvaluateFunctionTableDriven(t *testing.T) {
```
для ассертов - https://github.com/stretchr/testify#assert-package
Это упростит код

jskonst commented on golang/tests/functions_test.go:
```diff
@@ -0,0 +1,34 @@
+		a, b, x float64
+		want    string
+	}{
+		{0.4, 0.8, 3.2, "0.991503"},
```
правило, которого стоит придерживаться - 1 тест 1 кейс

EugeneWWolf replied:
Переделала тесты с помощью ассертов, но не уверена, соблюла ли я правило 1 тест 1 кейс. Скорее всего, мой код можно сделать чище и читабельнее.
Старалась написать как можно лучше, но мне нужен сон.

jskonst commented on golang/tests/functions_test.go:
```diff
@@ -0,0 +1,34 @@
+		{0.4, 0.8, 3.21, "0.990128"},
+	}
+
+	for _, testInput := range tests {
```
не надо вам все тестировать - 1 полложительный - а дальше - краевые случаи, даже если задачей у вас они не обозначены

jskonst commented on golang/main.go:
```diff
@@ -21,12 +13,12 @@ func main() {
 	step := 0.6
 	for x := startValueForX; x <= endValueForX; x += step {
-		printFunctionValue(x, evaluatingFunction(a, b, x))
+		internal.PrintFunctionValue(x, internal.EvaluateFunction(a, b, x))
```
Задачи А и Б вытаскивайте в отдельные функции
а вот тесты на них - только то что они размер массива правильный вернут

EugeneWWolf committed:
Minor fixes (changed module name back from "myProject" to "isuct.ru/informatics2022")

EugeneWWolf committed:
- Tried to simplify my tests (used asserts from "github.com/stretchr/testify/assert")
- Made SolveTaskA and SolveTaskB functions, now they solve task A and task B respectively
- Remade PrintFunctionValue (now it prints return values (slices) from both SolveTaskA and SolveTaskB), it puts each element of a slice on a new line
- Used pointers in some places to try to improve my code

EugeneWWolf committed:
Tried to make my code more universal, added user input, improved tests, added error handling

EugeneWWolf commented:
Постаралась сделать код более "универсальным", теперь он может принимать пользовательский ввод. Но код выглядит так, будто его надо отрефакторить позже.

EugeneWWolf committed:
Tried to simplify and improve my code (added type structs in functions_test.go, improved error messages, added new comments to a few places)

EugeneWWolf committed:
Added one more test to functions_test.go

jskonst commented on golang/internal/functions.go:
```diff
@@ -0,0 +1,56 @@
+	return functionValue, nil
+}
+
+func PrintFunctionValue(functionValue *[]float64) {
```
если у нас это срез то он уже указатель

jskonst commented on golang/main.go:
```diff
@@ -2,31 +2,46 @@ package main
-	step := 0.6
-	for x := startValueForX; x <= endValueForX; x += step {
-		printFunctionValue(x, evaluatingFunction(a, b, x))
+	var a, b float64
```
лучше сразу все же присваивайте значения

EugeneWWolf replied:
У пользователя есть возможность самостоятельно вводить начальное и конечное значение для переменной x, для параметров a и b, поэтому нельзя определить их сразу. В коде значения задаются через fmt.Scan и указатели.

jskonst replied:
не надо - в ближайшее время мы с пользовательским вводом не работаем
а если будет нужно - то эту логику работы с пользовательским вводом надо выносить отдельно

jskonst commented on golang/main.go:
```diff
@@ -2,31 +2,46 @@ package main
-		printFunctionValue(x, evaluatingFunction(a, b, x))
+	var a, b float64
+	var startValueForX, endValueForX, step float64
+	fmt.Println("Task A: please, input a, b, start value for x, end value for x and step for x") // Указатели сделаны, чтобы передавать в функции слайс
```
не совсем и можно было и не передавать как указатель - там копия только заголовка
https://stackoverflow.com/questions/39993688/are-slices-passed-by-value

jskonst commented on golang/tests/functions_test.go:
```diff
@@ -0,0 +1,86 @@
+func TestEvaluateFunction(t *testing.T) {
+	assert := assert.New(t)
+
+	testValues := testsEvaluateFunction{a: 0.4, b: 0.8, x: 3.2, want: "0.991503"}
```
не надо приводить к строке - потом используйте https://pkg.go.dev/github.com/stretchr/testify/assert#InDelta

jskonst commented on golang/tests/functions_test.go:
```diff
@@ -0,0 +1,86 @@
+	testValues := testsEvaluateFunction{a: 0.4, b: 0.8, x: 3.2, want: "0.991503"}
+
+	functionValue, err := internal.EvaluateFunction(testValues.a, testValues.b, testValues.x)
+	assert.Equal(testValues.want, fmt.Sprintf("%.6f", functionValue), "Want and got should be equal.")
```
```suggestion
	assert.Equal(testValues.want, fmt.Sprintf("%.6f", functionValue),  fmt.Sprintf("expected: %d got: %d", testValues.want, functionValue ))
```

jskonst commented on golang/tests/functions_test.go:
```diff
@@ -0,0 +1,86 @@
+	assert.Equal(testValues.want, fmt.Sprintf("%.6f", functionValue), "Want and got should be equal.")
+
+	testValues = testsEvaluateFunction{a: 0.0, b: 0.0, x: 0.0}
+	functionValue, err = internal.EvaluateFunction(testValues.a, testValues.b, testValues.x)
```
1 тест - проверка только 1 случая - это у вас еще + 2 функции тестов

jskonst commented on golang/tests/functions_test.go:
```diff
@@ -0,0 +1,86 @@
+	want       int
+}
+
+func TestSolveTaskB(t *testing.T) {
```
Для задач а и б достаточно проверить просто длины массивов + тесты на неправильные параметры

jskonst reviewed (changes requested):
Оч хорошо - но надо еще улучшить

EugeneWWolf committed:
- Removed unnecessary pointers at some places
- Used assert.InDelta to compare float64 values instead of Sprintf and float64 to string conversion
- Refactored functions_test.go (used 1 test 1 case rule)
- Removed some tests in function_test.go

jskonst commented on golang/internal/functions.go:
```diff
@@ -0,0 +1,56 @@
+func SolveTaskB(a, b float64, size int, variableValues []float64) []float64 {
+	sliceOfResults := make([]float64, 0, size)
+
+	for i := 0; i < size; i++ {
```
можно было for range

jskonst commented on golang/main.go:
```diff
@@ -2,31 +2,46 @@ package main
-	for x := startValueForX; x <= endValueForX; x += step {
-		printFunctionValue(x, evaluatingFunction(a, b, x))
+	var a, b float64
+	var startValueForX, endValueForX, step float64
```
по хорошему  - это надо выносить из main'а
кероме того - еще раз повторюсь - не надо пока нам
по факту сейчас редко работают вот так интерактивно - скорее через аргументы запуска, конфигарации, переменные окружения

EugeneWWolf committed:
- Made one flag (-d) for my program (without flag the program will use standart values of variables, otherwise (if you set the flag to false) the user will be able to pass their own values
- Created file readUserInput.go and wrote code to accept user input

EugeneWWolf commented:
Теперь программа при запуске будет работать с заранее заданными значениями. Но если пользователь хочет, то он может присвоить значение false флагу d (go run main.go -d=false) и он сможет интерактивно ввести свои произвольные значения.

(Я оставила интерактивный ввод, так как переписывать прогу практически полностью долго, но если будет время, я попытаюсь ради практики сделать произвольный ввод значений с помощью флагов)

**Переделала код (см. ниже)**

EugeneWWolf committed:
Refactored my code, now it works with command line arguments

EugeneWWolf committed:
Minor improvements

EugeneWWolf committed:
Solved all codewars tasks

jskonst commented on golang/go.mod:
```diff
@@ -1,3 +1,5 @@
 
 go 1.16
+
+require github.com/stretchr/testify v1.8.1 // indirect
```
👍
